src/perform_step/adams_bashforth_moulton_perform_step.jl

```diff
@@ -1608,32 +1607,86 @@ function perform_step!(integrator, cache::ABCN2Cache, repeat_step=false)
   cnt = integrator.iter
   f1 = integrator.f.f1
   f2 = integrator.f.f2
+  f1(du₁, uprev, p, t)
+  @. k1 = integrator.fsalfirst - du₁
+  # Explicit part
   if cnt == 1
-    @. u = uprev + dt*integrator.fsalfirst
-    f2(k2, uprev, p, t)
+    @. tmp = uprev + dt * k1
   else
-    f1(du₁, uprev, p, t)
-    # Explicit part
-    @. k1 = integrator.fsalfirst - du₁
     @. tmp = uprev + dt * (1.5*k1 - 0.5*k2)
-    cache.k2 .= k1
-    # Implicit part
-    # mass_matrix = integrator.sol.prob.mass_matrix
+  end
+  # Implicit part
+  # precalculations
+  γ = 1//2
+  γdt = γ*dt
+  new_W = calc_W!(integrator, cache, γdt, repeat_step)
+
+  # initial guess
+  @. z = dt*du₁
+  @. tmp += γ*z
+  z,η,iter,fail_convergence = diffeq_nlsolve!(integrator, cache, W, z, tmp, γ, 1, Val{:newton}, new_W)
+  fail_convergence && return
+  @. u = tmp + 1//2*z
+
+  cache.k2 .= k1
+  cache.ηold = η
+  cache.newton_iters = iter
+  integrator.f(k,u,p,t+dt)
+end
 
-    # precalculations
-    γ = 1//2
-    γdt = γ*dt
-    new_W = calc_W!(integrator, cache, γdt, repeat_step)
+# CNLF2
 
-    # initial guess
-    @. z = dt*du₁
-    @. tmp += γ*z
-    z,η,iter,fail_convergence = diffeq_nlsolve!(integrator, cache, W, z, tmp, γ, 1, Val{:newton}, new_W)
-    fail_convergence && return
-    @. u = tmp + 1//2*z
+function initialize!(integrator,cache::CNLF2ConstantCache)
+  integrator.kshortsize = 2
+  integrator.k = typeof(integrator.k)(integrator.kshortsize)
+  integrator.fsalfirst =
+  integrator.f.f1(integrator.uprev,integrator.p,integrator.t) +
+  integrator.f.f2(integrator.uprev,integrator.p,integrator.t) # Pre-start fsal
+
+  # Avoid undefined entries if k is an array of arrays
+  integrator.fsallast = zero(integrator.fsalfirst)
+  integrator.k[1] = integrator.fsalfirst
+  integrator.k[2] = integrator.fsallast
+end
 
-    cache.ηold = η
-    cache.newton_iters = iter
+function perform_step!(integrator,cache::CNLF2ConstantCache,repeat_step=false)
+  @unpack t,dt,uprev,u,f,p = integrator
+  @unpack k2,uprev2,uf,κ,tol = cache
+  cnt = integrator.iter
+  f1 = integrator.f.f1
+  f2 = integrator.f.f2
+  du₁ = f1(uprev, p, t)
+  # Explicit part
+  if cnt == 1
+    tmp = uprev + dt * (integrator.fsalfirst - du₁)
+  else
+    tmp = uprev2 + 2 * dt * (integrator.fsalfirst - du₁)
   end
-  integrator.f(k,u,p,t+dt)
+  # Implicit part
+  # precalculations
+  if cnt == 1
+    γ = 1
```

Comment:
this isn't type-stable. Make it `1//1`. There's probably a similar change that should be done in other places.